### PR TITLE
refactor(sbb-dialog): introduce inner components and header animation

### LINKED ISSUE #2019: [Bug]: SBB-Dialog Content is not scrollable / accessible
### Preflight Checklist

- [X] I have read the [Contributing Guidelines](https://github.com/lyne-design-system/lyne-components/blob/master/CONTRIBUTING.md) for this project.
- [X] I agree to follow the [Code of Conduct](https://github.com/lyne-design-system/lyne-components/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://github.com/orgs/lyne-design-system/projects/4) for an issue that matches the one I want to file, without success.

### Bug type

Accessibility

### It affects the following packages

Components

### Which version of Lyne Components are you using

0.40.14

### Operating system

_No response_

### Browser / Browser version

_No response_

### Input mode

None

### Additional settings

_No response_

### What happened?

With the high font magnification (400%), the pages are not scrollable, so the application cannot be operated.
zoom

### Steps to reproduce the issue

1. Step one
2. Step two
3. ...

### Relevant log output

_No response_

### PR DESCRIPTION
**Pull request by dauriamarco**
Closes #2019 

BREAKING CHANGE: The `sbb-dialog` component now supports the dedicated inner elements `sbb-dialog-title`, `sbb-dialog-content`, and `sbb-dialog-actions`. Use these components to respectively provide a title, a content and, optionally, a footer with an action group. Moreover, the full-screen variant (which occurred when no title was provided to the dialog) has been removed. To achieve a full-screen overlay, please use the `sbb-overlay` component.
This was the previous implementation:
```html
<sbb-dialog title-content="Title">
  <p>Dialog content.</p>
  <sbb-action-group slot="action-group">...</sbb-action-group>
</sbb-dialog>
```
This is the new implementation:
```html
<sbb-dialog>
  <sbb-dialog-title>Title</sbb-dialog-title>
  <sbb-dialog-content>Dialog content</sbb-dialog-content>
  <sbb-dialog-actions>...</sbb-dialog-actions>
</sbb-dialog>
```
Previously, a full-screen dialog was displayed if no title was provided to the dialog component:
```html
<sbb-dialog>
  <p>Dialog content.</p>
</sbb-dialog>
```
It is now possible to use an `sbb-overlay` component to display the same variant:
```html
<sbb-overlay>
  <p>Overlay content.</p>
</sbb-overlay>
```

### CODE REVIEW

**Review by jeripeierSBB — CHANGES_REQUESTED** [review]
Thanks for addressing this!
Code looks ok so far, but testing it in real life, the initial scrolling looks buggy for me as it forces the user to scroll down at least the height of the header. Is this really intended?

Please also wait for approval of @mcilurzo

**Comment by dauriamarco** [comment]
> Thanks for addressing this! Code looks ok so far, but testing it in real life, the initial scrolling looks buggy for me as it forces the user to scroll down at least the height of the header. Is this really intended?
> 
> Please also wait for approval of @mcilurzo

@jeripeierSBB Yes, I know what you mean, but I have already shown it to Manuel and it is fine with him. But if you want we can show it to him again and discuss it, although I can't think of any other possible alternatives at the moment.

**Comment by jeripeierSBB** [comment]
We would like to discuss with @mcilurzo if there is the need of a property to opt out of the header hiding feature. potentially providing a breakpoint range from where it stops or starts... to be discussed.

**Comment by kyubisation** [comment]
We discussed this with @mcilurzo.
We would like to have additional elements `sbb-dialog-title` (extends from `sbb-title`), `sbb-dialog-content` and `sbb-dialog-actions` (extends from `sbb-action-group`).
`sbb-dialog-title` should have a property `hideOnScroll: false | '' | 'zero' | 'micro' | ... = false;`, which indicates up to (including) which breakpoint to title should be hidden when scrolled (`''` loosely equals to 'ultra', in order for `<sbb-dialog-title hide-on-scroll>` to work).

File: src/components/dialog/dialog-content/readme.md
Comment (DavideMininni-Fincons) [inline src/components/dialog/dialog-content/readme.md:5]: Maybe add an example with the dialog?
```
<sbb-dialog params>
  <sbb-dialog-content>Dialog content</sbb-dialog-content>
</sbb-dialog>
```

File: src/components/dialog/dialog-title/readme.md
Comment (DavideMininni-Fincons) [inline src/components/dialog/dialog-title/readme.md:4]: See other similar comment to wrap it in a sbb-dialog tag

File: src/components/dialog/dialog-title/readme.md
Comment (DavideMininni-Fincons) [inline src/components/dialog/dialog-title/readme.md:17]: Add examples (see title's readme as comparison)

File: src/components/dialog/dialog/readme.md
Comment (DavideMininni-Fincons) [inline src/components/dialog/dialog/readme.md:71]: Add an example here

File: src/components/dialog/dialog-actions/readme.md
Comment (DavideMininni-Fincons) [inline src/components/dialog/dialog-actions/readme.md:4]: See other comments about wrapping

**Review by DavideMininni-Fincons — CHANGES_REQUESTED** [review]
Great work. I like the class extension (from title and action-group).

File: src/components/dialog/dialog/dialog.ts
Comment (TomMenga) [inline src/components/dialog/dialog/dialog.ts:221]: Not sure we need this

File: src/components/dialog/dialog/dialog.ts
Comment (TomMenga) [inline src/components/dialog/dialog/dialog.ts:136]: While I understand the other inits, why do we need to fetch this element each open?

**Review by TomMenga — CHANGES_REQUESTED** [review]
Good job! 👍 
Some minor discussion points and fixes

File: src/components/dialog/dialog-title/dialog-title.ts
Comment (jeripeierSBB) [reply src/components/dialog/dialog-title/dialog-title.ts:38]: As the whole change is a breaking change, I think we could just rename it. A quick lookup in internal code database showed me that it anyways isn't used at all.

File: src/components/dialog/dialog/dialog.stories.ts
Comment (TomMenga) [inline src/components/dialog/dialog/dialog.stories.ts:73]:
```suggestion
  options: ['off', 'zero', 'micro', 'small', 'medium', 'large', 'wide', 'ultra'],
```
Should we add an 'off' value?
And in the story title template something like `hide-on-scroll=${hideOnScroll !== 'off' ? hideOnScroll : nothing}`

**Review by TomMenga — CHANGES_REQUESTED** [review]
All good, just a nitpick in stories

File: src/components/dialog/dialog/dialog.stories.ts
Comment (dauriamarco) [reply src/components/dialog/dialog/dialog.stories.ts:73]: I see your point but I'm afraid it'd be confusing for the consumer who might think that `off` would be a possible value for the property.

Comment (TomMenga) [reply src/components/dialog/dialog/dialog.stories.ts:73]: true... can't see a good solution that makes everyone happy :/

**Review by TomMenga — APPROVED** [review]

File: src/components/dialog/dialog/dialog.ts
Comment (jeripeierSBB) [inline src/components/dialog/dialog/dialog.ts:406]: Don't we exclude close button with that?

File: src/components/dialog/dialog/dialog.ts
Comment (jeripeierSBB) [inline src/components/dialog/dialog/dialog.ts:390]: we would have a field for sbbdialogtitle

**Review by jeripeierSBB — CHANGES_REQUESTED** [review]
Thanks for the huge refactoring!
Some points to do and to discuss are left. The coupling of all the component is very tight (css var / data-attributes and logic), but probably it's okay because they are only meant to be used together.

- Do we need to include testA11yTreeSnapshot() in all specs? @TomMenga 
- [ ] Check HCM
-  [ ] Mention the breaking changes in PR description (including removed fullscreen support)

**Comment by TomMenga** [comment]
> * Do we need to include testA11yTreeSnapshot() in all specs? @TomMenga

As a general rule, yes. At least one `testA11yTreeSnapshot` of the most meaningful scenario for each component.

For @dauriamarco, we have 3 possible ways to implement them:
- When the template is taken care of by a beforeEach (see `accordion.spec.ts`)
- When the template is elementary (see `clock.spec.ts`)
- Otherwise, we can convert a spec test to a describe with a beforeEach (see `expansion-panel.spec.ts`)

If you have any doubts let me know 😸

File: src/components/dialog/dialog/dialog.ts
Comment (dauriamarco) [reply src/components/dialog/dialog/dialog.ts:406]: In this case not anymore, as we are now slotting the close button with the `sbb-dialog-title`.

File: src/components/dialog/dialog-title/dialog-title.scss
Comment (dauriamarco) [reply src/components/dialog/dialog-title/dialog-title.scss:48]: I asked Manuel and he confirmed no shadow on negative.

File: src/components/dialog/dialog/dialog.ts
Comment (dauriamarco) [reply src/components/dialog/dialog/dialog.ts:239]: I added this because we would see a flash on the first open if it's negative, so I had to sync it earlier.

File: src/components/dialog/dialog/dialog.ts
Comment (dauriamarco) [reply src/components/dialog/dialog/dialog.ts:240]: I added this because we would see a flash on the first open if it's negative, so I had to sync it earlier.

Comment (jeripeierSBB) [reply src/components/dialog/dialog/dialog.ts:240]: ok. can you add a comment there why this is there?

File: src/components/dialog/dialog/dialog.ts
Comment (dauriamarco) [reply src/components/dialog/dialog/dialog.ts:239]: I need this to sync the negative state before the first open. I could remove it from `open()`, even though the title might change on the next opens and keeping it can be more stable (?) You tell me 👍

Comment (jeripeierSBB) [reply src/components/dialog/dialog/dialog.ts:239]: just keep it and add a comment :)

File: src/components/dialog/dialog/dialog.ts
Comment (jeripeierSBB) [reply src/components/dialog/dialog/dialog.ts:406]: Ah I see because the focusable function goes into the shadow DOM of sub components.

File: src/components/dialog/dialog/dialog.ts
Comment (dauriamarco) [reply src/components/dialog/dialog/dialog.ts:131]: It breaks only if the users don't provide a title or a content. Those are mandatory and that's why I used the "!", I always expect to have them both. The only optional element is the `sbb-dialog-actions`. If the user provides something else it will be simply not be shown.

File: src/components/dialog/dialog/dialog.ts
Comment (dauriamarco) [reply src/components/dialog/dialog/dialog.ts:133]: Not sure, if I use the host all the overflow and scrolling logic would break.

Comment (jeripeierSBB) [reply src/components/dialog/dialog/dialog.ts:133]: do we have any advantage of using a separate component for the content?

File: src/components/dialog/dialog/dialog.ts
Comment (jeripeierSBB) [reply src/components/dialog/dialog/dialog.ts:131]: Ok, but open is never called there, sorry.

File: src/components/dialog/dialog/dialog.ts
Comment (jeripeierSBB) [inline src/components/dialog/dialog/dialog.ts:135]: Should not better this._dialogActionsElement be able to deal with null?
```suggestion
    this._dialogActionsElement = this.querySelector('sbb-dialog-actions');
```

File: src/components/dialog/dialog-title/readme.md
Comment (jeripeierSBB) [inline src/components/dialog/dialog-title/readme.md:5]: in the case you provide an accessibility-back-label string you should also display backButton ?

**Comment by mcilurzo** [comment]
looks good to me!

File: src/components/dialog/dialog-title/readme.md
Comment (dauriamarco) [reply src/components/dialog/dialog-title/readme.md:5]: Not sure, maybe a question for @kyubisation?

Comment (kyubisation) [reply src/components/dialog/dialog-title/readme.md:5]: Good question. Might be an implicit opt-in.
Would make sense to me, but we should mention this in the `backButton` documentation.

File: src/components/dialog/dialog/readme.md
Comment (jeripeierSBB) [inline src/components/dialog/dialog/readme.md:14]: I'd never explicitely write the slot in the examples as it is assigned automatically

File: src/storybook/pages/home/home--logged-in.stories.ts
Comment (jeripeierSBB) [inline src/storybook/pages/home/home--logged-in.stories.ts:214]: Also in the examples I'd avoid to set the slot explicitly
```suggestion
```

**Review by jeripeierSBB — CHANGES_REQUESTED** [review]

**Review by jeripeierSBB — APPROVED** [review]
LGTM, thanks!!

**Pull request merged**
